docs/rfcs/20220413-aot-for-all-snode.md

```diff
@@ -0,0 +1,363 @@
+  ti.root.dense(ti.ij).place(f)
+```
+
+This introduces confusion at several spots:
+
+1. Type is not purely decided by `dtype`, but also by how the field is placed.
+2. It introduces the notion of "nested field", which Taichi doesn't currently have a good abstraction for. Because of this, it is quite complicated to apply certain kind of optimizations for a composite-typed field. For example, vectorized load/save consumes the same bandwidth as scalar ops on certain platforms. Without a good abstraction, the checking for whether a matrix field is AoS or SoA has to be spread across different passes in CHI IR.
+
+If we further think about the problem, SoA `x` is not really a field. Instead, it is a *grouped view* of three individual scalar fields. This view provides matrix operations, which won't make sense for each individual scalar field.
+
+In addition, because type is currently coupled with Taichi field definition, a Taichi field has to be implemented as individual fields in order to support the SoA scenario. Once we switch to the type builder pattern, we can control how the type is constructed first, and choose the field implementation later.
+
+If we want to make it explicit that this is a *field view*, we can do the following:
+
+```py
+builder = ti.SNodeTreeBuilder()
+builder.add_field(dtype=ti.f32, name='v0')
+builder.add_field(dtype=ti.f32, name='v1')
+builder.add_field(dtype=ti.f32, name='v2')
+for v in ['v0', 'v1', 'v2']:
+  builder.tree().dense(ti.ij, 4).place(v)
+
+# Checks that
+# 1. `components` and `dtype` are compatible.
+# 2. All the fields in `components` are homogeneous in their SNode hierarchy.
+# 3. `components` themselves are not field views.
+builder.add_field_view(dtype=ti.vec3, name='vel', components=['v0', 'v1', 'v2'])
+```
+
+Matrix field view supports common matrix operations, and is equivalent to expanding each component into a local matrix variable.
+
+```py
+# 1
+vel_soa[i, j].inverse()
+# equivalent to
+ti.vec3([v0[i, j], v1[i, j], v2[i, j]]).inverse()
+
+# 2
+vel_soa[i, j][1] += 2.0
+# equivalent to
+v1[i, j] += 2.0
+
+# 3
+vel_soa[i, j] = vel_soa[i, j] @ some_vec3
+# equivalent to
+vel_tmp = ti.vec3([v0[i, j], v1[i, j], v2[i, j]])
+vel_tmp = vel_tmp @ some_vec3
+v0[i, j] = vel_tmp[0]
+v1[i, j] = vel_tmp[1]
+v2[i, j] = vel_tmp[2]
+```
+
+To make field view even more powerful, we can supported *nested field views*. For example:
+
+```py
+vertex_t = ti.types.struct({'pos': ti.vec3, 'normal': ti.vec3})
+sphere_t = ti.types.struct({'center': vertex_t, 'radius'; ti.f32})
+
+builder = ti.SNodeTreeBuilder()
+builder.add_field(dtype=ti.vec3, name='pos')
+builder.add_field(dtype=ti.vec3, name='normal')
+builder.add_field(dtype=ti.f32, name='radius')
+builder.add_field_view(dtype=sphere_t, name='spheres', 
+                       compoenents=[['pos', 'normal'], 'radius'])
+###                                 ^^^^^^^^^^^^^^^^^ Note this is nested
+```
+
+## Gradient and AutoDiff
```

Comment:
Added discussion on gradient fields @ailzhang @strongoier

Reply:
Are users going to manually place the grad fields? The new way is much cleaner, but I have a feeling that it is not possible to get autodiff backward-compatible..

Reply:
@strongoier could you elaborate on what'll be bc-breaking? ;) We currently allow manually placing grad fields as well? https://github.com/taichi-dev/taichi/blob/master/python/taichi/examples/autodiff/simple_derivative.py#L10

Reply:
Echo what @ailzhang has said. This basically works like before:

1. If you invoke `add_field(dtype, name='x', needs_grad=True, shape=(a, b, c))`. Then you shouldn't `place('x')` anymore. The builder will just generate two new dense SNodes. (This is actually talking about two things. If you define `shape` but not `needs_grad`, the builder will still do so, but just one dense SNode)
2. Otherwise you need to manually place the field and its gradient field?

Reply:
In current Taichi, the grad field is always created. `needs_grad` only helps you place it when `shape` is specified. This is different from the description in this RFC, where `needs_grad` is used to determine whether a field has a grad field.

BTW there is another commonly used helper function `lazy_grad()` for placing, which should also be supported somewhere.

Reply:
Good point about `lazy_grad`! maybe that's something we can add to the snodetreebuilder? 

TIL we always create grad field but sometimes don't place it (nice catch!).... But I'd expect nothing should break if we only create the grad field when needs_grad=True or lazy_grad is on in current Taichi? (If so we should probably change our current implementation)

Reply:
Sounds good! We can have a `lazy_grad()` API for `SNodeTreeBuilder`. The implementation needs to be changed anyway when we switch to the new API, so I think the behavior that grad fields are always created can be altered then.

Reply:
Thanks for the discussion! Yeah, I didn't know that grad field is always created, either... I feel like we ought to correct this implementation.

Reply:
Fixed